libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -226,13 +203,26 @@ onError g m r k e = liftIO $ do
 
 -- | Log an 'Error' response for debugging purposes.
 logError :: MonadIO m => Logger -> Maybe Request -> Error -> m ()
-logError g r (Error c l m) = liftIO $ Log.debug g logMsg
+logError g mr (Error c l m) = liftIO $ Log.debug g logMsg
   where
     logMsg = field "code" (statusCode c)
            . field "label" l
-           . field "request" (fromMaybe "N/A" (lookupRequestId =<< r))
+           . field "request" (maybe "N/A" reqinfo mr)
            . msg (val "\"" +++ m +++ val "\"")
 
+    -- Using unsafePerformIO here is important because we only want it to run if log level is
+    -- set to 'Debug'.
+    reqinfo r = show ( lookupRequestId r
+                     , r
+                       -- show the bytestring containing the request body, this way it can be
+                       -- read back into ghci for forensics.
+                     , IOUnsafe.unsafePerformIO $ lazyRequestBody r
+
+                         -- TODO: not sure this will still find the request body intact?  it
```

Comment:
If this is a concern (perhaps unsafePerformIO doesn't guarantee the proper ordering in this case) 
then maybe we instead explicitly introspect the logLevel, e.g. `if (Log.level . Log.settings $ Logger) == Debug` or something similar and just do it all in IO? I guess it's not too scary since this would only cause problems if we're on Debug; but I wouldn't want my app to stop working just because we switched logging to Debug haha.

Reply:
I checked, and you can't run `lazyRequestBody` twice, because the first time sucks the stream empty and the second time will come up empty.  I'll think of something else.